rfcs/20200810-tf-internal-api.md

```diff
@@ -0,0 +1,119 @@
+related APIs. <b>(See alternative naming in the sections below, the naming of 
+the namespace will be discussed in the design meeting.)</b>
+
+### Backward compatibility
+The contract for all the APIs under this namespace is looser compared to core TF
+API. It will remain backward compatible for at least 1 minor TF release. 
+
+This means for any API that is added in 2.x release, it will remain the same in 
+2.x+1. If we choose to remove the same API, we will mark it as deprecated in 
+2.x+1, and delete it at 2.x+2 release. TensorFlow is released every 3 months, 
+and this will give enough time for all the clients to change to the new 
+API/alternative.
+
+Any deprecation and backward incompatible API change will be explicitly called 
+out in TF release notes. This applies to both core API and "internal" API.
+
+### Acceptance Criteria
+The candidate of the "internal" API should:
+
+1. Does NOT fit for core TF API, otherwise it should be exposed as core TF API.
+1. Are effectively required to build/test/deploy a framework or library on top 
+   of TensorFlow, or to integrate such a framework into TensorFlow's API.
+1. Mature enough and won't change behavior/interface for every release. 
+1. Non-trivial, otherwise the function should be copied to the client side.
+1. Widely used in the implementation of TF public APIs (i.e. new functionality 
+   isn’t immediately added to the tf.internal namespace)
+1. Has at least two downstream libraries which are known to need it.
+
+TF API owners will review the new API proposal, and follow the existing review 
+process for core TF API.
+
+
+### Documentation and Docstring
+The "internal" API should have the same style and standard as the core 
+TensorFlow API, which is documented [here](https://github.com/tensorflow/community/blob/master/governance/api-reviews.md#docstrings). 
+We should explicitly list out the difference between "internal" API and core
+API, and also choose a different place on tensorflow.org so that existing user 
+are not confused.
+
+### Naming and sub namespace
+Similar "internal" APIs should be grouped together as sub namespaces, e.g., test 
+related APIs should live under "tf.internal.test". This is aligned with the 
+existing TF naming conversion.
+
+Try not to export experimental APIs since the "internal" API should be mature 
+enough.
+
+### Current candidate
+The following list is created from the private TF method usage within Keras, 
+when we were trying to convert Keras to use the public TF API only. This is by 
+no means the full list, but will serve as the first wave of review requests we 
+send to the API owner for review. We don't expect all of them to be approved, 
+and will discuss with the API owner on a case to case basis.
+
+|Symbol location  |API Name  | 
+:-------------- |:---------------------------------------------------- |
+|python.framework.func_graph.FuncGraph |tf.internal.FuncGraph     |
```

Comment:
@saxenasaurabh mentioned usages could be replaced with `tf.function(f).get_concrete_function(inputs).graph`? That would be preferable if feasible.

Reply:
Would the `tf.function` traces get saved in some sort of never-cleared cache? Or is it possible to explicitly gc them? part of our usage of `FuncGraph` is to encapsulate traces that can then be immediately cleaned up.

Reply:
And also related to this: would it trace twice rather than once? That would also be a performance hit to our existing usages.

Reply:
I would much prefer exposing this type of behavior as regular arguments to get_concrete_function (or similar), since these do appear to be valid optimizations that other framework authors may want to use.

Reply:
It would be helpful to see the actual usages to determine this, but to answer the questions:

- Caching: Unless the graph is held elsewhere, deleting the instance of tf.function(f) should delete everything.
- Number of traces: Once if first trace doesn't create any variable, Twice if first trace creates variables

cc @saxenasaurabh to keep me honest.

Reply:
Don't we need to trim the FuncGraph API before considering exposing this?

Reply:
+1 to @kkimdev. Caching should not be a concern here because the lifetime of cached entries is tied to the instance returned by `tf.function(f)`.
Regarding multiple traces, it should not be a concern if variables are not being created. And if there are, then using `tf.function` is actually safer I think.